src/shogun/base/SGObject.cpp

```diff
@@ -975,12 +975,12 @@ std::string CSGObject::to_string() const
 	return ss.str();
 }
 
-std::vector<std::string> CSGObject::parameter_names() const
+std::map<std::string, std::reference_wrapper<const AnyParameter>> CSGObject::get_parameters() const
```

Comment:
I just realised that `get_parameters()` seems to be used a few times in shogun's code base, so will need to change this to something else. @karlnapf do you have any suggestions? `parameter_map()`?

Reply:
I’ll need to check where that is and whether it can be changed.
Maybe we’ll have a snowball effect of moving c code to swig

Reply:
I don't think there are many, there is one in neuralnets (and then DeepAutoencoders) and rbm, from what I can tell

Reply:
the std::map is not supported for all the languages we support in SWIG... that's why actually we have CMap in the codebase :( so either we have to go with our map or.... it's only std::vector that is being implemented in all of the language mappings

Reply:
Ah ok! I tried with CMap but I get an error because the copy constructor is implicitly deleted.. I have this
```
        CMap<std::string, std::string> parameter_types() const {
            CMap<std::string, std::string> result;
            for (auto const& each: self->get_parameters()) {
                result[each.first] = each.second.get().get_value().type();
            }
            return result;
        }
``` 
How do I get SWIG to copy CMap, when it wraps result with SwigValueWrapper?

Reply:
mmmm i would allocate it on the heap, and return a `CMap<...>*`

Reply:
so basically

```
CMap<std::string, std::string>* parameter_types() const {
            auto result = new CMap<std::string, std::string>();
            for (auto const& each: self->get_parameters()) {
                result->add(each.first, each.second.get().get_value().type());
            }
            return result;
        }
```

Reply:
Ah thanks! I got it to work :)

Reply:
ok so another thing is that maybe we wanna keep the parameter_names function and have an additional one that returns the map?

Reply:
Do you mean in libshogun or in SWIG?

Reply:
both... so that we still have just a function that returns the params :)

Reply:
we actually discussed that the parameter names is not part of any libshogun API and thus not really needed from C++ perspective (it used to be when we didnt have non-owning any policies).
So, following the idea that interface code should not part part of libshogun, I think we should have both as part of swig. And there, yes I think we should have both the map and the vector

Reply:
However, I am against using CMap as this thing should not even be there. Instead, we can have a getter for single parameter descriptions (this is fine with your intended use as well). 
Furthermore, we could then either have Python/specific lang extensions that turn it into a local datastructure (such as an explicit typemap for std::map in Python) or whatever if we want.

I just think it would be so awkward to actually get a CMap instance from the Python interface when asking for descriptions. Then rather have a list comprehension [obj.get_param_descr(name) for name in obj.parameter_names()] or something.

Oh and finally another solution could be to just put the descriptions as a vector in the same order as the parameter names.

All these solutions are ugly :(

Reply:
update.

Just saw you already wrote a cmap typemap for python, so I guess this already works well.

I am still a bit skeptical as this makes the interface between python and the other langs (that dont have the cmap typemap) different, which is something we usually want to avoid. Having a single getter avoids this problem at the cost of the loop necessary if users wanna get all descriptions .... Not that bad in interactive environments like Python.

At the end of the day, this is a question of taste I guess